File: three_pigs/three_pigs.py
from __future__ import print_function
import copy

class ThreePigs(object):
    def __init__(self):
        self.templates = [
            [ [[-1,0],[0,1]], [[0,-1],[1,0]], [[1,0],[0,-1]], [[0,1],[-1,0]] ],
            [ [[0,1,0]], [[0],[1],[0]] ],
            [ [[-1,-1,0],[0,0,1]], [[0,-1],[0,-1],[1,0]], [[1,0,0],[0,-1,-1]], [[0,1], [-1,0], [-1, 0]]],
        ]
        self.arrangements = []
        self._cal_all_arrangements()

    def _cal_all_arrangements(self):
        b = [
            [-1,0,0,-1],
            [0,0,0,0],
            [0,0,0,0],
            [0,0,0,-1]
        ]
        self._search(b, 0, [])

    def _search(self, b, step, trace):
        for p in self.templates[step]:
            matches = self._match(b, p)
            for m in matches:
                new_b = self._apply(b, p, m)
                new_trace = trace.copy()
                new_trace.append((p,m))
                if step == 2:
                    self.arrangements.append((new_trace, new_b))
                else:
                    self._search(new_b, step+1, new_trace)

    def _apply(self, b, p, m):
        h = len(p)
        w = len(p[0])
        i,j = m
        new_b = copy.deepcopy(b)
        for di in range(h):
            for dj in range(w):
                if p[di][dj] == 0:
                    new_b[i+di][j+dj] = 1
                if p[di][dj] == 1:
                    new_b[i+di][j+dj] = 2
        return new_b

    def _match(self, b, p):
        matches = []
        h = len(p)
        w = len(p[0])
        for i in range(4-h+1):
            for j in range(4-w+1):
                matched = True
                for di in range(h):
                    for dj in range(w):
                        if p[di][dj] != -1 and b[i+di][j+dj] != 0:
                            matched = False
                if matched:
                    matches.append((i,j))
        return matches

File: three_pigs/three_pigs.js
class ThreePigs {
  constructor() {
    this.templates = [
      [ [[-1,0],[0,1]], [[0,-1],[1,0]], [[1,0],[0,-1]], [[0,1],[-1,0]] ],
      [ [[0,1,0]], [[0],[1],[0]] ],
      [ [[-1,-1,0],[0,0,1]], [[0,-1],[0,-1],[1,0]], [[1,0,0],[0,-1,-1]], [[0,1], [-1,0], [-1, 0]]],
    ];
    this.arrangements = [];
    this._cal_all_arrangements();
  }

  _cal_all_arrangements() {
    var b = [ [-1,0,0,-1], [0,0,0,0], [0,0,0,0], [0,0,0,-1] ];
    this._search(b, 0, []);
  }

  _search(b, step, trace) {
    for (let p of this.templates[step]) {
      var matches = this._match(b, p);
      for (let m of matches) {
        var new_b = this._apply(b, p, m);
        var new_trace = [...trace];
        new_trace.push([p,m]);
        if (step == 2) {
          this.arrangements.push([new_trace, new_b]);
        } else {
          this._search(new_b, step+1, new_trace);
        }
      }
    }
  }

  _deepcopy(o) {
    return JSON.parse(JSON.stringify(o));
  }

  _apply(b, p, m) {
    var h = p.length;
    var w = p[0].length;
    var i = m[0];
    var j = m[1];
    var new_b = this._deepcopy(b);
    for (let di = 0; di < h; di++) {
      for (let dj = 0; dj < w; dj++) {
        if (p[di][dj] == 0) {
          new_b[i+di][j+dj] = 1;
        }
        if (p[di][dj] == 1) {
          new_b[i+di][j+dj] = 2;
        }
      }
    }
    return new_b;
  }

  _match(b, p) {
    var matches = [];
    var h = p.length;
    var w = p[0].length;
    for (let i = 0; i < 5-h; i++) {
      for (let j = 0; j < 5-w; j++) {
        var matched = true;
        for (let di = 0; di < h; di++) {
          for (let dj = 0; dj < w; dj++) {
            if (p[di][dj] != -1 && b[i+di][j+dj] != 0) {
              matched = false;
            }
          }
        }

        if (matched) {
          matches.push([i, j]);
        }
      }
    }
    return matches;
  }
}

module.exports = {
  ThreePigs
}
